### DISCUSSION TITLE
Consider Allowing Optional Parameters for Inferred Lambdas

### DISCUSSION THREAD

**Discussion by JasonBock** [dotnet/csharplang#5421] [Language Ideas]
C# 10 added inferred lambdas, so we could define lambdas with a little less text:
```
var noOptionals = (string a, int b) => a.Length + b;
```

We can also add attributes to the lambda if we want:
```
var noOptionals = ([Description("first parameter")] string a, int b) => a.Length + b;
```

Furthermore, we can declare parameters to be `ref` if we want:
```
var withRefs = (ref string a) => a.Length;
```

This last one surprised me when I saw it demonstrated at [.NET Conf 2021](https://youtu.be/dfzBMxXQUOc?t=985), as a `Func<,>` can't be used in this case. The compiler generates a delegate for this situation - it looks something like this:
```
[CompilerGenerated]
internal delegate TResult <>F{00000001}<T1, TResult>(ref T1);
```

Inferred lambdas will also generate a delegate if you declare more than 16 parameters, something that can't be done with a `Func` or `Action` (IMO declaring a method with that many parameters isn't a good idea, but it works):
```
var aLotOfParameters = (int x0, int x1, int x2, int x3,
  int x4, int x5, int x6, int x7, int x8,
  int x9, int x10, int x11, int x12,
  int x13, int x14, int x15, int x16, int x17, int x18) =>
  x0 + x1 + x2 + x3 + x4 + x5 + x6 + x7 +
    x8 + x9 + x10 + x11 + x12 + x13 + x14 + x15 + x16 + x17 + x18;
```

However, currently optional arguments are not supported with inferred lambdas, even though they can be expressed in a delegate. For example, I can't do this:

```
var optionals = (string a, int b = 2) => a.Length + b;
```

Even though I can write a delegate like this:
```
public delegate int optionalsDelegate(string a, int b = 2);
```

I don't have any specific use cases where I need this functionality myself, though it seems like it would be natural to support this case when attributes, many parameters, and `ref` parameters are supported for inferred lambdas.

Side Note: There may be other cases that aren't supported with inferred lambdas. For example, it doesn't seem like you can use generics:
```
public void Foo<T>()
{
  // This is OK
  var withGenerics = new Func<T>(() => default);
  // This doesn't work.
  var withGenericsInferred = <T>() => default(T);
}
```

That's purely a guess on the syntax - I don't know if it generics **are** supported, but I know what I tried doesn't work, though you can declare a delegate for it:
```
public delegate T withGenericDelegate<T>();
```

And if you did support generics, then do you allow constraints as well?
```
var withGenerics = <T>(T a) where T : ISomeInterface => a.SomeMethodDefinedOnISomeInterface();
```
I'm guessing this has been discussed already :). Just bringing up the point that there may lots of cases that aren't supported with inferred lambdas, not just optional parameters.

**Comment by HaloFour** [comment]
I think the question would have to be under what use cases would those features be useful.  It's not enough to want to achieve a "completeness", each has to justify the work to support the syntax changes given they can't be expressed that way.

**Reply by JasonBock** [reply]
@HaloFour correct, that is something I forgot to state in the original post. There's always the cost of justifying the work, is it necessary, etc. I was just a bit surprised that `ref` was supported, but there had to have been justification for that to be included.

**Comment by jnm2** [comment]
Since return types can now be specified, here's the syntax that does work:

```cs
void Foo<T>()
{
    // C# 9
    var withGenerics = new Func<T>(() => default);
    // C# 10
    var withGenericsInferred = T () => default;
}
```
